### PR TITLE
[WIP] Use Express-style path matching

### PR DESCRIPTION
**Pull request by taion**
This is a remake of https://github.com/reactjs/react-router/pull/3105.

This is not quite ready for release, as I want to make this an opt-in feature guarded behind a deprecation for now, as well as to update the documentation and examples as appropriate.

This PR changes the pattern matching algorithm from the home-grown React Router to `path-to-regexp`, which is used by Express and a number of other JavaScript routing frameworks.

The biggest advantages of this change are:
- No more need to maintain a bespoke pattern matching algorithm, and instead use a syntax consistent with the rest of the JavaScript ecosystem
- Support for regex-qualified route parameters, for constraining route patterns (https://github.com/reactjs/react-router/issues/2286)
- Dramatically increased performance, though for reasons somewhat unclear to me (https://github.com/reactjs/react-router/issues/3215, providing a more optimal implementation than is available on React Router)

This will come with a few breaking changes once the feature is enabled:
- Optional segments must be specified as `/:foo?` as with Express, rather than `(/:foo)` as with RR v1+ (but this is the syntax in v0.13.x)
- Splats now come through as numbered parameters, rather than under `splat`
- Repeated named parameters will take the last value, rather than collapsing into an array
- Non-greedy splats are no longer supported; `*` is a greedy splat

In other words, Express's path matching works slightly differently to React Router's.

I expect that for the majority of applications, this should be fully drop-in. For the rest, the migration work will be relatively limited.

### CODE REVIEW

**Comment by taion** [comment]
And for what it's worth, I don't like the Express syntax for optional segments, but I feel like consistency here trumps my aesthetic preferences.

**Comment by taion** [comment]
My current thoughts for backward compatibility strategy:
- Allow setting `patternMatcher={version}` on a route-by-route basis
- Allow a top-level global static default flag for changing the pattern matcher version
- In production mode, use the specified pattern matcher
- In development mode, if the old pattern matcher is specified, run both pattern matchers, and warn if the results differ
